rust/catalyst-types/src/id_uri/role_index.rs

```diff
@@ -12,54 +12,135 @@ use thiserror::Error;
 /// Role Index parsing error
 #[derive(Display, Error, Debug)]
 #[allow(clippy::module_name_repetitions)]
-pub enum RoleIndexError {
+pub enum RoleIdError {
+    /// Invalid parse
+    Parse(#[from] ParseIntError),
+
     /// Invalid Role Index
-    InvalidRole(#[from] ParseIntError),
+    InvalidRole(u8),
 }
 
 /// Project Catalyst User Role Index.
 ///
 /// <https://github.com/input-output-hk/catalyst-CIPs/blob/x509-catalyst-role-definitions/CIP-XXXX/README.md>
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RoleIndex(u16);
+#[repr(u8)]
+#[non_exhaustive]
+pub enum RoleId {
+    /// Primary required role use for voting and commenting.
+    Role0 = 0,
+
+    /// Delegated representative (dRep) that vote on behalf of delegators.
+    DelegatedRepresentative = 1,
 
-impl RoleIndex {
-    /// Delegated Representative
-    pub const DREP: RoleIndex = RoleIndex(1);
-    /// Proposer
-    pub const PROPOSER: RoleIndex = RoleIndex(3);
-    /// Default Role 0
-    pub const ROLE_0: RoleIndex = RoleIndex(0);
+    /// Proposer that enabling creation, collaboration, and submission of proposals.
+    Proposer = 3,
+
+    /// A custom role.
+    Custom(u8),
+}
 
-    /// Is the `RoleIndex` the default value (Role 0)
+impl RoleId {
+    /// Is the `RoleId` the default value [`Self::Role0`]
     #[must_use]
     pub fn is_default(self) -> bool {
-        self == Self::ROLE_0
+        self == Self::Role0
+    }
+
+    /// Returns the `u8` representation of the role.
+    #[must_use]
+    pub const fn as_u8(self) -> u8 {
+        match self {
+            RoleId::Role0 => 0,
+            RoleId::DelegatedRepresentative => 1,
+            RoleId::Proposer => 3,
+            RoleId::Custom(b) => b,
+        }
+    }
+
+    /// Returns `true` if the role belongs to the canonical set of known roles.
+    #[must_use]
+    pub const fn is_known(self) -> bool {
```

Comment:
Better replace this with using this derive macro https://docs.rs/strum/latest/strum/derive.EnumIs.html

Reply:
No no, I've meant to not have an explicit `is_known` method defined by ourselves at all.
rather to use `EnumIs` so it will define for us something like `is_unknown` which would be more than enough to call `!is_unknown` in any other place.